File: sdk/typescript/tools/live-acceptance.mjs
import assert from 'node:assert/strict';
import {createHash, randomUUID} from 'node:crypto';
import process from 'node:process';

const packageName = '@capsem/sdk';
/** @type {unknown} */
const built = await import(packageName);
const {Hypervisor, VmLifecycleState} = /** @type {typeof import('../src/index.js')} */ (built);
const url = process.env.SDK_GATEWAY_URL, token = process.env.SDK_GATEWAY_TOKEN;
assert(url && token, 'Fixture gateway URL and token are required');
/**
 * @param {import('../src/index.js').ExecOutput} output
 * @param {string} expected
 */
function assertOutput(output, expected) {
  assert.equal(output.encoding, 'utf8');
  assert.equal(output.data, expected);
}
const hv = new Hypervisor(url, token, {timeoutMs: 120_000});
const name = `sdk-ts-${randomUUID().slice(0, 8)}`;
try {
  const requestedProfile = process.env.CAPSEM_TEST_PROFILE;
  const profile = requestedProfile === undefined
    ? undefined
    : (await hv.profiles.list()).find(item => item.id === requestedProfile);
  assert(requestedProfile === undefined || profile !== undefined, `missing profile ${requestedProfile}`);
  const vm = await hv.create({name, cpus: 2, memory: 2, ...(profile === undefined ? {} : {profile})});
  assertOutput((await vm.exec('printf SDK_EXEC_READY')).stdout, 'SDK_EXEC_READY');
  const info = await vm.info();
  assert.equal(info.id, vm.id);
  assert.equal(info.status, VmLifecycleState.RUNNING);
  assert(info.persistent && info.ai && info.network && info.files);
  const data = Uint8Array.from({length: 4364}, (_, index) => index % 256);
  const upload = await vm.files.write('sdk-proof.bin', data);
  assert(upload.success && upload.size === data.length);
  assert.deepEqual(await vm.files.read('sdk-proof.bin'), data);
  assert((await vm.files.list()).entries.some(entry => entry.name === 'sdk-proof.bin'));
  const digest = createHash('sha256').update(data).digest('hex');
  const exec = await vm.exec('sha256sum /root/sdk-proof.bin; printf SDK_STDERR >&2; exit 7');
  assert.equal(exec.exit_code, 7);
  assert.equal(exec.stdout.data.split(' ')[0], digest);
  assertOutput(exec.stderr, 'SDK_STDERR');
  await vm.log({tail: 10});
  await hv.log({tail: 10});
  await vm.history({limit: 10});
  await vm.stats.summary();
  await vm.stats.details();
  await vm.snapshots.list();
  await vm.snapshots.status();
  const stopped = await vm.stop();
  assert(stopped.success && stopped.persistent);
  assert.equal((await vm.info()).status, VmLifecycleState.STOPPED);
  const fork = await vm.fork(`${name}-fork`);
  assert.notEqual(fork.id, vm.id);
  assert.equal((await fork.info()).forked_from, vm.id);
  await fork.start();
  assert.deepEqual(await fork.files.read('sdk-proof.bin'), data);
  await fork.files.write('sdk-proof.bin', new Uint8Array([42]));
  await vm.start();
  assert.deepEqual(await vm.files.read('sdk-proof.bin'), data);
  await vm.pause();
  assert.equal((await vm.info()).status, VmLifecycleState.SUSPENDED);
  await vm.resume();
  assertOutput((await vm.exec('printf SDK_RESUMED')).stdout, 'SDK_RESUMED');
  assert.deepEqual(await vm.files.read('sdk-proof.bin'), data);
  // On failure the owning service fixture preserves evidence before cleanup.
  await fork.delete();
  await vm.delete();
  assert(!(await hv.list()).sandboxes.some(entry => entry.name === name || entry.name === `${name}-fork`));
  process.stdout.write(`SDK_LIVE_ACCEPTANCE_OK bytes=${String(data.length)} sha256=${digest}\n`);
} finally {hv.close();}

File: sdk/python/tests/live_acceptance.py
"""Exercise public SDK lifecycle and byte transfer against a disposable real VM."""

from __future__ import annotations

import asyncio
import hashlib
import os
import uuid

from capsem import VM, Hypervisor, models


async def ready(vm: VM) -> None:
    async with asyncio.timeout(60):
        while True:
            result = await vm.exec("printf SDK_EXEC_READY", timeout_secs=5)
            if result.exit_code == 0 and result.stdout == "SDK_EXEC_READY":
                return
            await asyncio.sleep(0.2)


async def main() -> None:
    async with Hypervisor(os.environ["SDK_GATEWAY_URL"], os.environ["SDK_GATEWAY_TOKEN"], timeout=120) as hv:
        name = "sdk-live-" + uuid.uuid4().hex[:8]
        vm = await hv.create(os.environ.get("CAPSEM_TEST_PROFILE", "code"), name=name, vcpu=2, memory="2G")
        fork: VM | None = None
        completed = False
        try:
            await ready(vm)
            info = await vm.info()
            assert info.id == vm.id and info.name == name and info.persistent
            assert info.status == models.VmLifecycleState.RUNNING
            assert info.ai is not None and info.network is not None and info.files is not None
            assert info.cpus == 2 and info.ram_mb == 2048
            data = bytes(range(256)) * 17 + b"\x00SDK_BINARY\xff"
            uploaded = await vm.copy.to_vm("sdk-proof.bin", data)
            assert uploaded.success and uploaded.size == len(data)
            assert await vm.copy.from_vm("sdk-proof.bin") == data
            assert "sdk-proof.bin" in {entry.name for entry in (await vm.list()).entries}
            digest = hashlib.sha256(data).hexdigest()
            executed = await vm.exec("sha256sum /root/sdk-proof.bin; printf SDK_STDERR >&2; exit 7")
            assert executed.exit_code == 7 and executed.stdout.split()[0] == digest
            # The guest exec channel currently combines stdout and stderr.
            assert executed.stdout.endswith("\nSDK_STDERR") and executed.stderr == ""
            assert isinstance(await vm.log(tail=10), models.LogsResponse)
            assert isinstance(await hv.log(tail=10), models.HostLogsResponse)
            assert isinstance(await vm.history(limit=10), models.HistoryResponse)
            assert isinstance(await vm.stats.summary(), models.VmStatsSummaryResponse)
            assert isinstance(await vm.stats.details(), models.VmStatsDetailResponse)
            assert isinstance(await vm.snapshots.list(), models.SnapshotsList)
            assert isinstance(await vm.snapshots.status(), models.SnapshotsStatus)
            stopped = await vm.stop()
            assert stopped.success and stopped.persistent
            assert (await vm.info()).status == models.VmLifecycleState.STOPPED
            fork = await vm.fork(name + "-fork")
            assert fork.id != vm.id and (await fork.info()).forked_from == vm.id
            await fork.start()
            await ready(fork)
            assert await fork.copy.from_vm("sdk-proof.bin") == data
            await fork.copy.to_vm("sdk-proof.bin", b"fork-only")
            await vm.start()
            await ready(vm)
            assert await vm.copy.from_vm("sdk-proof.bin") == data
            await vm.pause()
            assert (await vm.info()).status == models.VmLifecycleState.SUSPENDED
            await vm.resume()
            await ready(vm)
            assert await vm.copy.from_vm("sdk-proof.bin") == data
            print(f"SDK_LIVE_ACCEPTANCE_OK bytes={len(data)} sha256={digest}")
            completed = True
        finally:
            # On failure the owning service fixture preserves evidence before
            # deleting its VMs; deleting here would erase the boot failure log.
            if completed:
                if fork is not None:
                    await fork.delete()
                await vm.delete()
        assert not any(entry.name in {name, name + "-fork"} for entry in (await hv.list()).sandboxes)


if __name__ == "__main__":
    asyncio.run(main())
